File: fs/problem_493.fs
(*
  $ fsharpc memoize.fs problem_493.fs
  $ mono problem_493.exe
*)

#nowarn "40"

let numColors = 7
let numPerColor = 10
let numToPick = 20

(* Determines if we've picked all the balls we need to from an urn. *)
let allBallsPicked = Seq.sum >> (=) (numColors * numPerColor - numToPick)

(* Counts the number of distinct colors that have been picked from the urn. *)
let countColorsPicked = Seq.filter ((<>) numPerColor) >> Seq.length

(* Returns a new urn with a ball of the given color removed. *)
let pick color urn =
  urn 
  |> List.indexed 
  |> List.map(fun (i, v) -> if (i = color) then v - 1 else v)

(* Tuple support. *)
let sumTuple (a1, a2) (b1, b2) = a1 + b1, a2 + b2

(* Computes (totalColorsPicked, totalLeaves) for leaves rooted at this subtree. *)
let rec computeCounts = Memoize.memoizeWithCacheKey List.sort (fun urn ->
  (* If this is a leaf, then we can just count the colors directly. *)
  if allBallsPicked urn then (bigint (countColorsPicked urn), 1I)
  (* Otherwise we need to sum up the values of all leaves rooted at this subtree. *)  
  else 
    [0..numColors-1] 
    |> Seq.map(fun color -> 
      [1..(List.item color urn)] 
      |> Seq.map(fun _ -> computeCounts (pick color urn)))
    |> Seq.concat
    |> Seq.reduce sumTuple
)

let startingUrn = List.replicate numColors numPerColor 
let totalColorsPicked, totalLeaves = computeCounts startingUrn
printfn "Total colors = %A" totalColorsPicked
printfn "Total leaves = %A" totalLeaves
printfn "Average = %0.9f" (float totalColorsPicked / float totalLeaves)

File: clj/problem_493.clj
(ns problem-493)

(def NUM_COLORS 7)
(def NUM_PER_COLOR 10)
(def NUM_TO_PICK 20)
(def SUM_ALL_PICKED (- (* NUM_COLORS NUM_PER_COLOR) NUM_TO_PICK))

;; Determines if we've picked all the balls we need to from an urn.
(defn all-balls-picked [urn]
  (= (reduce + urn) SUM_ALL_PICKED))

;; Counts the number of distinct colors that have been picked from the urn.
(defn count-colors-picked [urn]
  (bigint (count (filter 
    (fn [color] (< color NUM_PER_COLOR))
    urn))))

;; Returns the urn state after the given color is picked.
(defn pick [urn color] 
  (map-indexed 
    (fn [idx, count] 
      (if (= idx color) 
        (dec count) 
        count))
    urn))

;; Tuple support.
(defn sum-tuples [tuples]
  (reduce #(map + %1 %2) tuples))

;; Computes (total-colors-picked, total-leaves) for leaves rooted at this subtree.
(def urn-stats 
  (memoize (fn [urn]
    (if (all-balls-picked urn)
      ;; If this is a leaf, then we can just count the colors directly.
      (list (count-colors-picked urn) (bigint 1)),
      ;; Otherwise we need to sum up the values of all leaves rooted at this subtree.
      (sum-tuples 
        (mapcat 
          (fn [color] 
            (map 
              (fn [_] (urn-stats (sort (pick urn color))) )  ;; sort to get more cache hits
              (range 0 (nth urn color))))
          (range 0 NUM_COLORS)))))))

(let [starting-urn (repeat NUM_COLORS NUM_PER_COLOR)
      counts (urn-stats starting-urn)]
  (printf "Total colors = %s\n" (first counts))
  (printf "Total leaves = %s\n" (second counts))
  (printf "Average = %s\n" (double (reduce / counts))))
